src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5360,6 +5370,24 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         if (!(context.unit.isJava && cls.isClass)) NoSymbol
         else context.javaFindMember(cls.typeOfThis, name, s => s.isStaticMember || s.isStaticModule)._2
 
+      // If they try C.tupled, make it (C.apply _).tupled
+      def fixUpCaseTupled(tree: Tree, qual: Tree, name: Name, mode: Mode): Tree =
+        if (qual.symbol != null && currentRun.isScala3Cross && qual.symbol.isModule && qual.symbol.companion.isCase &&
+            context.undetparams.isEmpty && fixableFunctionMembers.contains(name)) {
+          val t2 = {
+            val t = atPos(tree.pos)(Select(qual, nme.apply))
+            val t1 = typedSelect(t, qual, nme.apply)
+            typed(atPos(tree.pos)(Select(etaExpand(t1, context.owner), name)), mode, pt)
+          }
+          if (!t2.isErroneous) {
```

Comment:
I was going to suggest using `silent` above, but actually the error message is helpful.

```
➜ sandbox qs -Xsource:3-cross
scala> case class D(x: Int, y: Int, z: Int)
scala> Some(1 -> 1).map(D.tupled)
                          ^
       error: type mismatch;
        found   : ((Int, Int, Int)) => D
        required: ((Int, Int)) => ?
```

Reply:
added a test